lib/segment/src/vector_storage/mmap_vectors.rs

```diff
@@ -118,4 +140,138 @@ impl MmapVectors {
+    ///
+    /// This is only supported on Unix.
+    fn lock_deleted_flags(&mut self) {
+        #[cfg(unix)]
+        if let Err(err) = self._deleted_mmap.lock() {
+            log::error!(
+                "Failed to lock deleted flags for quantized mmap segment in memory: {}",
+                err,
+            );
+        }
+    }
+}
+
+fn open_read(path: &Path) -> OperationResult<Mmap> {
+    let file = OpenOptions::new()
+        .read(true)
+        .write(false)
+        .append(true)
+        .create(true)
+        .open(path)?;
+
+    let mmap = unsafe { MmapOptions::new().map(&file)? };
+    madvise::madvise(&mmap, madvise::get_global())?;
+    Ok(mmap)
+}
+
+fn open_write(path: &Path) -> OperationResult<MmapMut> {
+    let file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(false)
+        .open(path)?;
+
+    let mmap = unsafe { MmapMut::map_mut(&file)? };
+    madvise::madvise(&mmap, madvise::get_global())?;
+    Ok(mmap)
+}
+
+/// Ensure the given mmap file exists and is the given size
+///
+/// # Arguments
+/// * `path`: path of the file.
+/// * `header`: header to set when the file is newly created.
+/// * `size`: set the file size in bytes, filled with zeroes.
+fn ensure_mmap_file_size(path: &Path, header: &[u8], size: Option<u64>) -> OperationResult<()> {
+    // If it exists, only set the length
+    if path.exists() {
+        if let Some(size) = size {
+            let file = OpenOptions::new().write(true).open(path)?;
+            file.set_len(size)?;
+        }
+        return Ok(());
+    }
+
+    // Create file, and make it the correct size
+    let mut file = File::create(path)?;
+    file.write_all(header)?;
+    if let Some(size) = size {
+        if size > header.len() as u64 {
+            file.set_len(size)?;
+        }
+    }
+    Ok(())
+}
+
+/// Get start position of flags `BitSlice` in deleted mmap.
+#[inline]
+const fn deleted_mmap_data_start() -> usize {
+    let align = mem::align_of::<usize>();
+    div_ceil(HEADER_SIZE, align) * align
+}
+
+/// Calculate size for deleted mmap to hold the given number of vectors.
+///
+/// The mmap will hold a file header and an aligned `BitSlice`.
+fn deleted_mmap_size(num: usize) -> usize {
+    let unit_size = mem::size_of::<usize>();
+    let num_bytes = div_ceil(num, 8);
+    let num_usizes = div_ceil(num_bytes, unit_size);
+    let data_size = num_usizes * unit_size;
+    deleted_mmap_data_start() + data_size
+}
+
+/// Create a convenient [`BitSlice`] view over a [`MmapMut`].
+///
+/// This works because the internal memory mapped slice is pinned and doesn't move in memory.
+///
+/// This is unsafe because we create a shared mutable refrence with lifetime `'a`.
+///
+/// - The mmap and BitSlice should never be mutated together.
+/// - The bitslice reference should never outlive the mmap.
+/// - The caller is responsible for handling this with care.
+unsafe fn mmap_to_bitslice<'a>(mmap: &mut MmapMut) -> &'a mut BitSlice {
```

Comment:
@IvanPleshkov @generall Maybe look at this in a bit more detail to make sure it is sound.